src/components/connection_handler/src/connection.cc

```diff
@@ -97,52 +98,78 @@ Connection::~Connection() {
   heart_beat_monitor_thread_->join();
   delete heartbeat_monitor_;
   threads::DeleteThread(heart_beat_monitor_thread_);
-  sync_primitives::AutoLock lock(session_map_lock_);
-  session_map_.clear();
-}
 
-// Finds a key not presented in std::map<unsigned char, T>
-// Returns 0 if that key not found
-namespace {
-template <class T>
-uint32_t findGap(const std::map<unsigned char, T>& map) {
-  for (uint32_t i = 1; i <= UCHAR_MAX; ++i) {
-    if (map.find(i) == map.end()) {
-      return i;
-    }
+  // Before clearing out the session_map_, we must remove all sessions
+  // associated with this Connection from the SessionConnectionMap.
+
+  // NESTED LOCK: make sure to lock session_map_lock_ then ConnectionHandler's
+  // session_connection_map_lock_ptr_ (which will be taken in RemoveSession).
+  sync_primitives::AutoLock lock(session_map_lock_);
+  SessionMap::iterator session_it = session_map_.begin();
+  while (session_it != session_map_.end()) {
+    LOG4CXX_INFO(
+        logger_,
+        "Removed Session ID "
+            << static_cast<int>(session_it->first)
+            << " from Session/Connection Map in Connection Destructor");
+    connection_handler_->RemoveSession(session_it->first);
+    session_it++;
   }
-  return 0;
+
+  session_map_.clear();
 }
-}  // namespace
 
-uint32_t Connection::AddNewSession() {
+uint32_t Connection::AddNewSession(
+    const transport_manager::ConnectionUID connection_handle) {
   LOG4CXX_AUTO_TRACE(logger_);
+
+  // Even though we have our own SessionMap, we use the Connection Handler's
+  // SessionConnectionMap to generate a session ID. We want to make sure that
+  // session IDs are globally unique, and not only unique within a Connection.
```

Comment:
Seems this comment should be moved just above the `AddSession` line.

Reply:
fixed in commit d4159c1768e28dccdd6027a714de59f93e06c8b2